File: guidebot_recorder/selects/visibility.js
/**
 * (el) => {visible, markerClass, enhanced} — how a `<select>` presents itself.
 *
 * The single answer to "is this control still the viewer's, or has the page
 * taken it over?". Four modules used to answer it four different ways, and the
 * drift between them is what produced two regressions on this branch: a
 * select2 original clipped to 1x1 px was "the control the viewer sees" to
 * validation while the shim called it enhanced and invisible, and a select
 * carrying a marker class was diagnosed purely off geometry, so the message
 * never mentioned the class that actually caused it.
 *
 * A bare expression on purpose, not an IIFE or a module: the JS side installs
 * it once as `window.__guidebot_select_shape` (see `Selects._script`) while the
 * Python side evaluates this very source inline — the recorder is handed a page,
 * not the controller that installed the widget, and `config.selects.mode:
 * native` installs no widget at all. Both therefore need an answer that does
 * not depend on the shim being present, and it has to be the same answer.
 *
 * `visible` is the geometric half: what "the page hid this control" means.
 * `markerClass` is the belt-and-braces half — the class name, not just a
 * boolean, so an error message can name it. The geometric test is the primary
 * signal because it is library-agnostic (select2 clips the original to 1x1 px,
 * Tom Select uses display:none, Chosen hides it too); the marker classes catch
 * the rest. `enhanced` is the disjunction, and it is the question every caller
 * outside this file actually asks.
 */
(el) => {
  const MARKER_CLASSES = ["select2-hidden-accessible", "tomselected", "chosen-select"];
  const computed = window.getComputedStyle(el);
  const rect = el.getBoundingClientRect();
  // 8x8 rather than "any box at all": a control the viewer is meant to click
  // is never smaller than that, and select2's 1x1 px original must not read as
  // on screen just because Playwright's `is_visible()` would say so.
  const visible = !(
    computed.display === "none" ||
    computed.visibility === "hidden" ||
    rect.width < 8 ||
    rect.height < 8
  );
  let markerClass = null;
  for (const name of MARKER_CLASSES) {
    if (el.classList.contains(name)) {
      markerClass = name;
      break;
    }
  }
  return {
    visible: visible,
    markerClass: markerClass,
    enhanced: !visible || markerClass !== null,
  };
}

File: guidebot_recorder/selects/visibility.py
"""The one predicate for "is this ``<select>`` already enhanced?".

``visibility.js`` holds the rule; this module is the thin Python accessor. Three
consumers read it and no consumer restates it:

* ``selects.js``'s ``isEnhanced`` — through the ``__guidebot_select_shape``
  global that :class:`guidebot_recorder.selects.selects.Selects` installs ahead
  of the widget body;
* ``recorder/select/_js.py``'s ``_SHIM_STATE_JS`` — by embedding :data:`SELECT_SHAPE_JS`
  into its own classification pass;
* ``resolver/widget.py``'s ``user_visible_control`` — through
  :func:`select_shape`.

Keeping the source in one file rather than the answer in one runtime object is
deliberate: the recorder and the validator both run in contexts where no widget
was installed (``config.selects.mode: native``, a health probe, a unit-test
page), so they cannot depend on the global — but they must not disagree with it.
"""

from __future__ import annotations

from importlib.resources import files
from typing import TypedDict

from playwright.async_api import Locator

#: ``(el) => {visible, markerClass, enhanced}`` — see ``visibility.js``.
#:
#: A JavaScript *expression*, so it drops straight into a Playwright
#: ``evaluate`` and can be parenthesised into a larger page function.
SELECT_SHAPE_JS = (
    files("guidebot_recorder.selects").joinpath("visibility.js").read_text(encoding="utf-8")
)

#: Where :class:`Selects` parks :data:`SELECT_SHAPE_JS` for ``selects.js``.
SELECT_SHAPE_GLOBAL = "__guidebot_select_shape"


class SelectShape(TypedDict):
    """What :data:`SELECT_SHAPE_JS` returns for one ``<select>``."""

    #: the control still has a clickable box of its own (the geometric half)
    visible: bool
    #: the marker class a widget library left behind, or ``None``
    markerClass: str | None
    #: ``not visible`` or a marker class — "the page took this control over"
    enhanced: bool


def shape_prelude() -> str:
    """The statement that publishes the predicate to page scripts."""

    return f"window.{SELECT_SHAPE_GLOBAL} = {SELECT_SHAPE_JS};\n"


async def select_shape(locator: Locator) -> SelectShape:
    """Read :data:`SELECT_SHAPE_JS` for the locator's single matching element."""

    return await locator.evaluate(SELECT_SHAPE_JS)
